packages/next/src/server/base-server.ts

```diff
@@ -2823,37 +2823,51 @@ export default abstract class Server<
           throw new NoFallbackError()
         }
 
-        if (!isNextDataRequest) {
-          // Production already emitted the fallback as static HTML.
-          if (isProduction) {
-            const html = await this.getFallback(
-              locale ? `/${locale}${pathname}` : pathname
-            )
+        if (!isNextDataRequest && !isRSCRequest) {
+          const key = isProduction
+            ? isAppPath
+              ? pathname
+              : locale
+                ? `/${locale}${pathname}`
+                : pathname
+            : null
+
+          const response = await this.responseCache.get(
```

Comment:
nit: maybe call this `fallbackResponse` or something so it's clearer what this codepath is doing